### ISSUE TITLE
Does JAX support distributed calculating?

### ISSUE THREAD

**Issue by Dream7-Kim** [google/jax#1460]
My video card is NVIDIA RTX 2070max-q and the VRAM is 8G.
It's not enough to calculate the gradient of my own functions, but I really need to do it.

So I have two plans to do:
(1) first one is to run the code on better video cards with 16G VRAM(tesla v100).
(2) The second one is to distribute the calculation.But I find that the support for distributed
calculation in jax is not so good, and it's now on experiencing step, right?

And i want to know : whether jax supports nvlink or some third-party tools for distributed calculating...

**Comment by jekbradbury** [comment]
If you have multiple GPUs in one machine, JAX’s `pmap` and `soft_pmap` functions will use NCCL to perform collectives/communication between the GPUs. Do you have multiple GPUs? What computation do you want to distribute?

**Comment by Dream7-Kim** [comment]
Yes, I have two Nvidia GTX 1080 on one machine, and my task is to calculate the gradient of a function. Thanks for you help and I will try to use them.

**Comment by Dream7-Kim** [comment]
Very glad to see Parallelization on reference. On September I cannot find it on reference, but after several days I found that with your advice. Thanks for author's update...

**Comment by Dream7-Kim** [comment]
```python
import jax
from jax import jit
from jax import jacrev
import jax.numpy as np
import numpy as onp
import time
from jax import lax

print("\n\n", jax.local_device_count())
print(jax.devices(), "\n\n")

NUM_VAR = 10**5
NUM_ARG = 200

rng = onp.random.RandomState(0)  # NOTE: control RNG for reproducibility

var_onp = rng.rand(NUM_VAR)
var = np.asarray(var_onp) # allocating the variables to GRAM

def func(arg):
    divider = 0. # NOTE: I made these floats
    numerator = 0.

    def body_fun(carry, x):
        divider, numerator = carry
        temp = np.dot(x, var)
        temp1 = np.sin(temp)
        temp2 = np.cos(temp)

        divid = np.add(temp1, temp2)
        divid = np.power(divid, 2)
        divid = np.sum(divid)

        numer = np.add(temp1, temp2)
        numer = np.sum(numer)
        numer = np.power(numer, 2)
        numerator = np.add(numer, numerator)

        divider = np.add(divider, divid)

        new_carry = divider, numerator
        return new_carry, ()

    (divider, numerator), _ = lax.scan(body_fun, (divider, numerator), arg)
    
    divider = np.power(divider, 1/2)

    return np.log(np.divide(numerator, divider))

# res = jax.jit(jax.grad(func))
res = jax.jit(jax.grad(func))

# arg = jax.random.normal(jax.random.PRNGKey(1), (NUM_ARG, 0))
np.set_printoptions(precision=16)
arg = rng.rand(NUM_ARG)
real_args = np.asarray(arg)
print("... ... ...Start calculating Gradient... ... ...")
start = time.time()
print(res(real_args))
# print(jax.pmap(res, devices=jax.devices()[0:])(real_args))
end = time.time()
print("First Execution time: ", end-start, " s ... ... ...")
```

This is my code. I have to calculate the gradient of this function and the result would have a type of (200, 1). I cannot use `pmap` adequately for this problem. My VRAM is not enough with `NUM_VAR=10**6`. Can anyone have some solution for this trouble?

**Comment by hawkinsp** [comment]
JAX supports multiple GPUs, including GPUs distributed across a cluster (https://jax.readthedocs.io/en/latest/multi_process.html).

Closing. If the example above is still interesting to discuss, I would reopen it as a new Github discussion.
